Muse Dash: Add New Game

DeamonHunter opened the pull request:
## What is this fixing or adding?
Adds Muse Dash as a core game.

## New Game Checklist
### Credits
- DeamonHunter (Discord: deamonhunter)

### Links
Client + APWorld: https://github.com/DeamonHunter/ArchipelagoMuseDash/releases/
Discord thread: https://discord.com/channels/731205301247803413/1063302131978862602

### How was this tested?
Tested over a decent period of time with games ran with unsupported syncs and asyncs.

No dedicated beta tests have been performed, but due to the amount of testing within unsupported games, I haven't felt the need to.

Berserker66 commented on setup.py:
```diff
@@ -74,6 +74,7 @@
     "Links Awakening DX",
     "Super Metroid",
     "SMZ3",
+    "Muse Dash"
```
```suggestion
    "Muse Dash",
```

Jarno458 reviewed (commented):
Did a review on the code, seems fine

Zunawe commented on worlds/musedash/__init__.py:
```diff
@@ -0,0 +1,289 @@
+            self.location_count = minimum_location_count
+
+    # Todo: Update this to list[str] when python 3.8 stops being used
+    def pop_random_item(self, list: list) -> str:
```
Pretty minor, but I think it's probably cleaner to remove this method and do

```python
self.multiworld.random.shuffle(my_list)

my_list.pop()
...
my_list.pop()
```

DeamonHunter replied:
Ah. Thanks for noticing this. Should now be fixed.

Zunawe commented on worlds/musedash/docs/setup_en.md:
```diff
@@ -0,0 +1,36 @@
+
+## Installing the Archipelago mod to Muse Dash
+
+1. Download [MelonLoader.Installer.exe](https://github.com/LavaGang/MelonLoader/releases/tag/v0.6.1) and run it.
```
It might be a better idea to link to https://github.com/LavaGang/MelonLoader/releases or https://github.com/LavaGang/MelonLoader/releases/latest instead of a specific tag which may eventually be out of date. Or if for some reason v0.6.1 is specifically the version of the loader we want to use and not newer, that should be mentioned in the text.

DeamonHunter replied:
I've updated all the links to /latest. Should definitely help with confusion.

DeamonHunter commented:
That should be the last feature added to this PR unless something comes up.

More recent changes of note:
- Trap loading changed a little bit.
- Slightly changed the data file. (And a little bit on reading it.)
- Some doc changes.

kindasneaki commented on worlds/musedash/docs/setup_en.md:
```diff
@@ -0,0 +1,46 @@
+- Windows 8 or Newer.
+- Muse Dash: [Available on Steam](https://store.steampowered.com/app/774171/Muse_Dash/)
+  - \[Optional\] Just As Planned DLC: [Also Available on Steam](https://store.steampowered.com/app/1055810/Muse_Dash__Just_as_planned/)
+- Melon Loader: [GitHub](https://github.com/LavaGang/MelonLoader/releases/latest)
```
I'm not sure if you should add that you need multiple .NET Runtimes for Melon loader or a link to its requirements

Silvris replied:
I believe it should just be .NET 6 now for MelonLoader? Might be useful to include

DeamonHunter replied:
Good point.
- Automated installer requires .net 4.8 installed.
- Melon loader requires .net 6 for this game in particular.
I'll update it so that these are mentioned.

DeamonHunter replied:
I've added the needed .net frameworks. Please check out if it makes sense.

kindasneaki commented:
I ran a few tests to try and break things, and the only things I found was if the `DLC is turned off and Master difficulty is selected` there aren't enough songs in the pool to generate. I already posted this in discord but figured I should put it here.
Others things I noticed in the code but aren't extremely important are that there are a bunch of if statements in `__init.py__` with parenthesis that aren't necessary for python.

DeamonHunter commented:
I've updated generation such that even the worst settings possible won't cause it to crash. I've also added tests for these cases.

Some stats:
- Streamer Mode on + DLC off leaves only 25 songs left to choose from.
- With the worst possibly starting setting, this gives 10 starting songs. I've forced it to try and get 2 * starting songs + 1 at a minimum. (Or 21 songs at the worst.)
- Thus it should always be possible to generate.

DeamonHunter commented:
A player ended up finding a bug where included songs would add on top of the requested number of songs. (They ended up with a 200~ song game when requesting 50.) So the recent changes were to address the oversight with the included song system.

ThePhar commented on worlds/musedash/docs/setup_en.md:
```diff
@@ -0,0 +1,48 @@
+## Joining a MultiWorld Game
+
+1. Launch Muse Dash and get past the intro screen. Click on the button in the bottom right.
+2. Enter in the details for the archipelago game, such as the server address with port (i.e archipelago.gg:38381), username and password.
```
Small grammar tweak. i.e. == "in essence" / e.g. == "example"

```suggestion
2. Enter in the details for the archipelago game, such as the server address with port (e.g. archipelago.gg:38381), username and password.
```

ThePhar commented on worlds/musedash/test/__init__.py:
```diff
@@ -0,0 +1,6 @@
+from typing import cast
```
Import not used.

ThePhar commented on worlds/musedash/test/__init__.py:
```diff
@@ -0,0 +1,6 @@
+from typing import cast
+from test.TestBase import WorldTestBase
+from worlds.musedash import MuseDashWorld
```
Ditto import not used.

ThePhar commented on worlds/musedash/test/__init__.py:
```diff
@@ -0,0 +1,6 @@
+from test.TestBase import WorldTestBase
+from worlds.musedash import MuseDashWorld
+
+class MuseDashTestBase(WorldTestBase):
```
Should have an extra line for PEP8 2 blank lines.

```suggestion

class MuseDashTestBase(WorldTestBase):
```

ThePhar reviewed (changes requested):
Most of my complaints are some style nitpicks, but mostly looks good. We get this cleaned up a bit, it should be good to go. \:)

DeamonHunter commented on worlds/musedash/__init__.py:
I knew I missed something... Thanks for that.

DeamonHunter commented:
Thanks Phar! That should be most of the changes, including a couple more pep8 changes picked up by the flake8 action.
